### PR TITLE
Version 0.2.0:  Browse and resolve zeroconf services 

### PR DESCRIPTION
**Pull request by marcschier**

### CODE REVIEW

File: api/csharp/src/Microsoft.Azure.Devices.Proxy/Model/BrowseClient.cs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+                    HostName = _host.Host,
+                    Interface = response.Interface,
+                    Aliases = response.Properties
+                        .Where(p => p.Type == (uint)DnsRecordType.Simple)
+                        .Select(p => Encoding.UTF8.GetString(((Property<byte[]>)p).Value)).ToArray()
+                };
+        }
+
+        /// <summary>
+        /// Enumerating resolved hosts
```
Comment (marcschier) [inline api/csharp/src/Microsoft.Azure.Devices.Proxy/Model/BrowseClient.cs]: Copy paste

File: api/csharp/src/Microsoft.Azure.Devices.Proxy/Model/BrowseClient.cs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+            protected override FileEntry Yield(BrowseResponse response) =>
+                new FileEntry {
+                    FileName = response.Item.AsProxySocketAddress().Host,
+                    Interface = response.Interface,
+                    Info = ((Property<FileInfo>)response.Properties.First()).Value
+                };
+        }
+
+        /// <summary>
+        /// Create service browser
```
Comment (marcschier) [inline api/csharp/src/Microsoft.Azure.Devices.Proxy/Model/BrowseClient.cs]: Comment

File: api/csharp/src/Microsoft.Azure.Devices.Proxy/Model/BrowseProtocol.cs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    public class BrowseResponse : Serializable<BrowseResponse>, IEquatable<BrowseResponse> {
+
+        /// <summary>
+        /// Request id
+        /// </summary>
+        [DataMember(Name = "handle", Order = 1)]
+        public Reference Handle { get; set; }
+
+        //
+        // Type of request - service, address resolver, folders
```
Comment (marcschier) [inline api/csharp/src/Microsoft.Azure.Devices.Proxy/Model/BrowseProtocol.cs]: Comment

File: api/csharp/src/Microsoft.Azure.Devices.Proxy/Model/FileEntry.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    using System;
+
+    /// <summary>
+    /// Browse files in directory
+    /// </summary>
+    public interface IDirectoryBrowser :
+        IAsyncEnumerator<FileEntry>, IDisposable { }
+
+    /// <summary>
+    /// Container class for file information
```
Comment (marcschier) [inline api/csharp/src/Microsoft.Azure.Devices.Proxy/Model/FileEntry.cs]: Copy paste

File: api/csharp/src/Microsoft.Azure.Devices.Proxy/Model/FileEntry.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    /// </summary>
+    public class FileEntry : IEquatable<FileEntry> {
+
+        /// <summary>
+        /// Contains host name
```
Comment (marcschier) [inline api/csharp/src/Microsoft.Azure.Devices.Proxy/Model/FileEntry.cs]: Comments bogus

File: api/csharp/src/Microsoft.Azure.Devices.Proxy/Serializer/Json.cs [hunk 6]
```diff
@@ -179,5 +186,86 @@ class SocketAddressConverter : JsonConverter {
...
+                        Value = jsonObject.GetValue("property").ToObject<AddressInfo>(serializer)
+                    };
+                }
+                else if (type == (uint)PropertyType.InterfaceInfo) {
+                    return new Property<InterfaceInfo> {
+                        Type = type,
+                        Value = jsonObject.GetValue("property").ToObject<InterfaceInfo>(serializer)
+                    };
+                }
+                else if (type > (uint)DnsRecordType.Simple) {
```
Comment (marcschier) [inline api/csharp/src/Microsoft.Azure.Devices.Proxy/Serializer/Json.cs]: Should check for upper bound

File: api/csharp/src/Microsoft.Azure.Devices.Proxy/Serializer/MsgPack.cs [hunk 7]
```diff
@@ -279,17 +271,188 @@ internal class SocketAddressSerializer : Serializer<SocketAddress> {
...
+                result = new Property<AddressInfo>(type,
+                    await context.Get<AddressInfo>().ReadAsync(
+                        reader, context, ct).ConfigureAwait(false));
+            }
+            else if (type == (uint)PropertyType.InterfaceInfo) {
+                result = new Property<InterfaceInfo>(type,
+                    await context.Get<InterfaceInfo>().ReadAsync(
+                        reader, context, ct).ConfigureAwait(false));
+            }
+            else if (type > (uint)DnsRecordType.Simple) {
```
Comment (marcschier) [inline api/csharp/src/Microsoft.Azure.Devices.Proxy/Serializer/MsgPack.cs]: Check for upper bound

File: api/csharp/src/Microsoft.Azure.Devices.Proxy/Serializer/MsgPack.cs [hunk 8]
```diff
@@ -279,17 +271,188 @@ internal class SocketAddressSerializer : Serializer<SocketAddress> {
...
+                await context.Get<AddressInfo>().WriteAsync(writer,
+                    ((Property<AddressInfo>)property).Value, context,
+                        ct).ConfigureAwait(false);
+            }
+            else if (property.Type == (uint)PropertyType.InterfaceInfo) {
+                await context.Get<InterfaceInfo>().WriteAsync(writer,
+                    ((Property<InterfaceInfo>)property).Value, context,
+                        ct).ConfigureAwait(false);
+            }
+            else if (property.Type > (uint)DnsRecordType.Simple) {
```
Comment (marcschier) [inline api/csharp/src/Microsoft.Azure.Devices.Proxy/Serializer/MsgPack.cs]: Check for upper bound

File: api/csharp/src/Microsoft.Azure.Devices.Proxy/Socket/SocketAddress.cs [hunk 9]
```diff
@@ -605,4 +763,87 @@ public class SocketAddressCollection : SocketAddress, IEquatable<SocketAddressCo
...
+        public override string ToString() {
+            return RemoteAddress.ToString();
+        }
+
+        /// <summary>
+        /// Returns hash for efficient lookup in list
+        /// </summary>
+        /// <returns></returns>
+        public override int GetHashCode() {
+            return (RemoteAddress.GetHashCode() * 31) ^
```
Comment (mregen) [inline api/csharp/src/Microsoft.Azure.Devices.Proxy/Socket/SocketAddress.cs]: Why always * 31?

Comment (marcschier) [inline api/csharp/src/Microsoft.Azure.Devices.Proxy/Socket/SocketAddress.cs]: Pseudo hashing.  I need to replace this with something better eventually.  Will open a bug.

**Pull request merged**